### PR TITLE
Add tests for batching support

### LINKED ISSUE #29179: Test batched equivalence
# Problem

A common assumption in our codebase and discussion is that batching has little impact on the results. However, we don't test it.

If we add the test below on `ModelTesterMixin`, we'll see that it fails on many models (`py.test tests/models -k test_batching_support`). We can also see that it passes on some models, like gpt2 or llama.

<details>
  <summary>Test Proposal (click me)</summary>

  ```py
    def test_batching_support(self):
        """
        Tests that the model supports batching and that the output is the nearly the same for the same input in
        different batch sizes.
        (Why "nearly the same" not "exactly the same"? Batching uses different matmul shapes, which often leads to
        different results: https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535)
        """

        def recursive_check(batched_object, single_row_object):
            if isinstance(batched_object, (List, Tuple)):
                for batched_object_value, single_row_object_value in zip(batched_object, single_row_object):
                    recursive_check(batched_object_value, single_row_object_value)
            elif isinstance(batched_object, Dict):
                for batched_object_value, single_row_object_value in zip(
                    batched_object.values(), single_row_object.values()
                ):
                    recursive_check(batched_object_value, single_row_object_value)
            elif batched_object is None:
                return
            else:
                batched_row = batched_object[0:1]
                self.assertFalse(torch.isnan(batched_row).any(), "Batched output has `nan`!")
                self.assertFalse(torch.isinf(batched_row).any(), "Batched output has `inf`!")
                self.assertFalse(torch.isnan(single_row_object).any(), "Single row output has `nan`!")
                self.assertFalse(torch.isinf(single_row_object).any(), "Single row output has `inf`!")
                self.assertTrue(
                    torch.allclose(batched_row, single_row_object, atol=1e-5),
                    msg=(
                        "Batched and Single row outputs are not equal. Difference="
                        f"{torch.max(torch.abs(batched_row - single_row_object))}."
                    ),
                )

        config, batched_input = self.model_tester.prepare_config_and_inputs_for_common()

        for model_class in self.all_model_classes:
            # Test as many model outputs as possible
            config.output_hidden_states = True
            config.output_attentions = True

            model = model_class(config).to(torch_device).eval()
            model_batched_output = model(**batched_input)

            batch_size = batched_input[model.main_input_name].shape[0]
            single_row_input = {}
            for key, value in batched_input.items():
                if isinstance(value, torch.Tensor) and value.shape[0] == batch_size:
                    single_row_input[key] = value[0:1]
                else:
                    single_row_input[key] = value
            model_row_output = model(**single_row_input)

            for key in model_batched_output:
                try:
                    recursive_check(model_batched_output[key], model_row_output[key])
                # Augments exception with the failing key to help debugging, if something went wrong
                except AssertionError as e:
                    raise AssertionError(
                        f"{e}\nError in key={key}"
                    )
  ```
</details>

# Plan forward

Iterate on the test above, fixing issues with the test and models. This issue can be tagged as complete when we have a test for "batching has little impact on the results", running on all models. Individual model fixes can be merged in advance of the test.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes #29179 . Currently the tests are passing for all models, and batched and unbatched outputs have nearly same outputs.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?
@gante

### CODE REVIEW

**Comment by gante** [comment]
@zucchini-nlp Running your test script on my setup I get

```
tensor(0., device='cuda:0')
tensor(4.7684e-07, device='cuda:0')
```

Could it be due to different hardware and/or versions? 🤔 For context, I'm running with CUDA 12.1, `torch==2.3.0.dev20240208+cu121`, and an nvidia RTX3090

File: tests/test_modeling_common.py
Comment (gante) [inline tests/test_modeling_common.py:777]:
```suggestion
                elif isinstance(value, torch.Tensor) and value.shape[0] % batch_size == 0:
                    # e.g. musicgen has inputs of size (bs*codebooks). in most cases value.shape[0] == batch_size
                    single_batch_shape = value.shape[0] // batch_size
                    single_row_input[key] = value[:single_batch_shape]
```
the former is a special case of the latter (right?)

File: tests/test_modeling_common.py
Comment (gante) [inline tests/test_modeling_common.py:785]: Are there cases where the output is a `torch.Tensor`? Even with `config.return_dict = True`?

If so, we should remove the `key` argument in `recursive_check`, to make things easier (and check all models)

**Review by gante — COMMENTED** [review]
Thank you for working on this important test case 🔥

File: src/transformers/models/tvp/modeling_tvp.py
Comment (zucchini-nlp) [reply src/transformers/models/tvp/modeling_tvp.py:682]: in the tests it was not... I did not find why that would happen, decided to leave it the way it works

File: src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py
Comment (zucchini-nlp) [reply src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py]: my bad, forgot to get the loss back

File: tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py
Comment (zucchini-nlp) [reply tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py:350]: okay. as the PR is a draft I will try to look in more detail for a few models that always fail

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:777]: makes sense

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:744]: okay, I will also check the versions and hardware as example resnet passed the test for you

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:786]: Yes, FlavaImageCodebook and SpeechT5HifiGan both return a tensor (not tuple of tensors) and do not have options for `return_dict`. I can change it, if it there isn't any special purpose for having tensor outputs

**Comment by zucchini-nlp** [comment]
Using cosine distance to check cv/audio models helped to solve the issue. The way we check if the model is conv-based or not might not be the best option,  I will see if I can make it better. 

Now everything is passing the test, yay!

**Review by gante — APPROVED** [review]
The end result is very cool! Thank you for iterating on this 🔥

**Comment by gante** [comment]
Next steps:
- `make fixup` :D
- `LayoutLMv2ModelTest` seems to be failing. Perhaps skit it as well?

File: tests/models/timm_backbone/test_modeling_timm_backbone.py
Comment (zucchini-nlp) [reply tests/models/timm_backbone/test_modeling_timm_backbone.py:204]: I think it's okay for Timm, anyway it is used several times as a backbone in CV models

File: tests/test_modeling_common.py
Comment (gante) [reply tests/test_modeling_common.py:794]: Uhmmm perhaps we should convert tensors to dicts before this line, so we check all classes

```py
if not isinstance(model_batched_output, torch.Tensor):
    model_batched_output = {"model_output": model_batched_output}
    model_row_output = {"model_output": model_row_output}
for key in model_batched_output:
    recursive_check(model_batched_output[key], model_row_output[key], model_name, key)
```

**Comment by gante** [comment]
@amyeroberts: you're not on rotation, but this PR includes a new test that, to make CI green, @zucchini-nlp had to a) make minor CV model changes; b) add skips on CV models 🤗 

Let us know if you find anything wrong! If the skips are not supposed to be skips, then I'd like to ask to keep them as TODOs so as to be tended by someone with CV expertise 🙏

File: tests/test_modeling_common.py
Comment (amyeroberts) [inline tests/test_modeling_common.py:766]: It seems like an easier fix would be to unify the model testers to have the batch size as standard

File: tests/test_modeling_common.py
Comment (amyeroberts) [inline tests/test_modeling_common.py:724]: Under what cases do these happen? It seems like they shouldn't happen with a batched output but will result in the teat passing here

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this - this is a great addition to making sure our models are well behaved 💪 

Just a few small comments

File: tests/models/longformer/test_modeling_longformer.py
Comment (zucchini-nlp) [reply tests/models/longformer/test_modeling_longformer.py:391]: yeah, good idea

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:724]: this is to catch loss or similar tensors, since some models have set "return_loss=True" by default

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:766]: hmm, okay, I will try

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py]: done

File: tests/models/timm_backbone/test_modeling_timm_backbone.py
Comment (zucchini-nlp) [reply tests/models/timm_backbone/test_modeling_timm_backbone.py]: okay, removed attentions then

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py]: Done and checked for all models and testers, not only the current test. Also, I added an error raising if the batch size is not 14 for any future model tests

File: tests/models/layoutlmv2/test_modeling_layoutlmv2.py
Comment (zucchini-nlp) [reply tests/models/layoutlmv2/test_modeling_layoutlmv2.py:422]: yeah, maybe I should check but I couldn't set up detectron2 correctly which is required for the test. I will try later again

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py]: That is the largest size I found from existing tests. If it is too big for tests, I can try something like 8 and test again, that nothing breaks. The only constraint I saw was that some models need an even batch size

Comment (amyeroberts) [reply tests/test_modeling_common.py]: Two questions:
- Why are we inforcing here for all the model testers to have the same value?
- Why take the max instead of the min?

Comment (zucchini-nlp) [reply tests/test_modeling_common.py]: - If a new model test is added and it does not use the standard batch size, the model might fail on the `batching_equivalence` test due to incorrectly inferred batch size (which is now not inferred from tester, but fixed to 14). As an option, we can fall back to taking the first row as one-batch input, so raising error will not not needed. But for particular models, where each batch can consist of multiple rows, the tests will fail anyway if they use anything expect for the standard 14. I decided to force the standard size, so that we do not have to worry about various batch size causing this test to fail in the future.
I am open to remove enforcing, and the future tests in case of failing will adjust batch sizes. Anyway, now I realize that most models have a standard one-bacth=one-row format :) 

- The max was chose for no particular reason, since I needed to have some guidance on which unified batch size to use for the test. Also, I was not sure if some model's tests rely heavily on larger batch size, so the max seemed like the better choice than min or avg

Comment (amyeroberts) [reply tests/test_modeling_common.py]: Sorry, I still don't understand why is has to be a fixed batch size for all models for this test. 

> But for particular models, where each batch can consist of multiple rows, the tests will fail anyway if they use anything expect for the standard 14.

Why is this the case i.e. what is it about the multiple rows that's causing it to fail? Is it specifically a batch size of 14 where this is true? 

The only thing I think we should be enforcing is that batch_size > 1. 

For some models, if they have special circumstances, then we can set `batch_size=14` for them. If there's additional  special logic required for those models, then they should have their own implementation `test_batching_equivalence` added in their model's test file

Comment (zucchini-nlp) [reply tests/test_modeling_common.py]: > It seems like an easier fix would be to unify the model testers to have the batch size as standard 

Maybe I misunderstood your first suggestion. I thought you suggested to use a fixed batch size for all testers, i.e. using a fixed batch size for this test, consequently for all testers.

> Why is this the case i.e. what is it about the multiple rows that's causing it to fail? Is it specifically a batch size of 14 where this is true?

Regarding this, it does not fail now, neither it failed before the last commit. It has nothing to do which batch size we use, but rather if we can correctly infer the batch size from data/testers. 

It will not fail for any future test also, but only if we infer the correct batch size, because we need to dissect out one batch out of all fake-data. If we do not succeed to dissect exactly one batch, the hidden states will result having high differences, the test fails.

Before "standard" batch size of 14, this case was handled by trying to infer `bs` from testers, but now we don't infer. Therefore, I decided to enforce, so no-one spends time to find the reason of same error again in the future. As another option, I can add a comment, not error about this case.

I hope it is clearer, but if not we can talk through it in a call. I guess that would be faster :D

Comment (amyeroberts) [reply tests/test_modeling_common.py]: > Maybe I misunderstood your first suggestion. I thought you suggested to use a fixed batch size for all testers, i.e. using a fixed batch size for this test, consequently for all testers.

@zucchini-nlp Ah, OK, sorry, I wasn't clear. What I meant was, we can just make sure all model testers have `batch_size` available for all model testers rather than having to search at different levels for composite models

Comment (zucchini-nlp) [reply tests/test_modeling_common.py]: Oh, that makes more sense. Got it, I will revert the changes back

Comment (zucchini-nlp) [reply tests/test_modeling_common.py]: done, added bs for composite testers where missing :)

File: tests/models/layoutlmv2/test_modeling_layoutlmv2.py
Comment (zucchini-nlp) [reply tests/models/layoutlmv2/test_modeling_layoutlmv2.py]: okay, had to install from source. Removed skipping for this model tester, it passed the test

File: tests/test_modeling_common.py
Comment (amyeroberts) [inline tests/test_modeling_common.py:715]: nit - let's use the normal primitives than the ones from typing 


```suggestion
            if isinstance(batched_object, (list, tuple)):
```

File: tests/test_modeling_common.py
Comment (amyeroberts) [inline tests/test_modeling_common.py:718]:
```suggestion
            elif isinstance(batched_object, dict):
```

File: tests/test_modeling_common.py
Comment (amyeroberts) [reply tests/test_modeling_common.py:726]: Can you put a comment to explain that here ?

File: tests/test_modeling_common.py
Comment (amyeroberts) [inline tests/test_modeling_common.py:768]: I don't think we need this condition given the if/elif structure and L776

**Review by amyeroberts — APPROVED** [review]
Thanks for adding - great work! 💪

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py]: Done, fixed the cases when `batch_size == num_hidden_layers` in certain testers, so there's no need for extra check now. Removed this `if`

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:727]: okay, done

**Comment by gante** [comment]
@zucchini-nlp the change in batch size has made the whisper tests unhappy :D after that is cleared up, we can merge this PR 🤗

**Review by ArthurZucker — APPROVED** [review]
Very nice PR! Thanks 

FYI @ydshieh  😉

**Comment by zucchini-nlp** [comment]
This PR is now ready to be merged. An unexpected bug was found yesterday, but it's fixed already. I will also merge `main` to resolve conflicts 
@gante

**Comment by ydshieh** [comment]
Hi @zucchini-nlp Thank you for your work. Our CI reports 2 failing tests:

```
tests/models/mamba/test_modeling_mamba.py::MambaModelTest::test_batching_equivalence
(line 726)  AttributeError: 'MambaCache' object has no attribute 'dim'
```

and 

```
tests/models/seggpt/test_modeling_seggpt.py::SegGptModelTest::test_batching_equivalence
(line 765)  AssertionError: tensor(False, device='cuda:0') is not true : Batched and Single row outputs are not equal in SegGptModel for key=hidden_states. Difference=0.1510644555091858.
```

If you can take a look on these, that would be great 🙏 Thank you.

**Comment by gante** [comment]
@ydshieh (and @amyeroberts, since you're working on the test fetcher): in this case, the new failures arise because the branch merged in this PR, which adds a new mixin test, was out of sync with `main` with respect to new models. Is there any way we can prevent this without too much effort? 

I'm not seeing an automated way to do it. The only thing that comes to mind is to ask to rebase before merging this type of PR :D

**Comment by ydshieh** [comment]
@gante Yeah, I don't see an satisfying way neither.

One possibility is to change circleCI config file to rebase (on the fly) on the latest main during the CI runs.

**Comment by gante** [comment]
> One possibility is to change circleCI config file to rebase (on the fly) on the latest main during the CI runs.

@ydshieh That could be... interesting! Try to rebase with `main`, roll back if a conflict arises, then run tests. If an informative message is added in the report (e.g. "the tests were run after rebasing on main"), we might save ourselves a few situations like this :D

**Comment by amyeroberts** [comment]
> One possibility is to change circleCI config file to rebase (on the fly) on the latest main during the CI runs.

I'm not sure this is a great idea, for two reasons: 
* the code version being used to run tests is not the version on the branch, which can lead to all sorts of confusion
* It's pretty common in development to not rebase for a few commits. Even though _technically_ PRs should be opened when the branch is review ready, this isn't what happens. It's going to end up forcing people to constantly rebase and force push, even if there aren't meaningful commits being added. As someone who gets a notification for every commit being pushed, I'd really rather avoid this. 

To test this out, I'd rather have an optional feature, similar to the run all tests command we can have in commits e.g. run-with-base, than we can comment on github, or ask users to run asa final step.

**Comment by gante** [comment]
>  than we can comment on github

That would also be quite cool -- and could save some time from the core maintainers, if the other reviewers diligently test against `main` on wide-reaching changes before merging 🤗

**Comment by ydshieh** [comment]
Totally agree with @amyeroberts that's is why I say `I don't see an satisfying way`

**Pull request merged**
